classes/source/source_url.php

```diff
@@ -0,0 +1,223 @@
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * URL source.
+ *
+ * @package    tool_etl
+ * @copyright  2019 John Yao <johnyao@catalyst-au.net>
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+
+namespace tool_etl\source;
+
+use curl;
+use tool_etl\config_field;
+use tool_etl\data;
+use tool_etl\logger;
+
+
+defined('MOODLE_INTERNAL') || die;
+
+global $CFG;
+require_once($CFG->libdir . '/filelib.php');
+
+class source_url extends source_base
+{
+
+    /**
+     * Name of the source.
+     *
+     * @var string
+     */
+    protected $name = "URL";
+
+    /**
+     * Settings.
+     *
+     * @var array
+     */
+    protected $settings = array(
+        'address' => '',
+    );
+
+    /**
+     * A temp folder to save files.
+     *
+     * @var string
+     */
+    protected $filedir;
+
+    /**
+     * Date and time now.
+     *
+     * @var string
+     */
+    protected $now;
+
+    public function __construct(array $settings = array())
+    {
+        global $CFG;
+
+        parent::__construct($settings);
+
+        $this->now = date('YmdHis', time());
+
+        $this->filedir = $CFG->dataroot . DIRECTORY_SEPARATOR . $this->get_short_name();
+        check_dir_exists($this->filedir);
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function extract()
+    {
+        if ($this->is_available()) {
+            $this->data = new data([$this->get_file()]);
+        } else {
+            throw new \Exception($this->name . ' source is not available!');
+        }
+
+        return $this->data;
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function is_available()
+    {
+
+        if ($this->check_remote_url($this->get_url_path(), $this->get_local_file_path($this->get_url_path()))) {
+            return true;
+        }
+
+        $this->log('load_data', 'URL is not valid ' . $this->get_url_path(), logger::TYPE_ERROR);
+
+        return false;
+    }
+
+    /**
+     * Copy file from source.
+     *
+     * @param string $remotefile Remote file path.
+     * @param string $localfile Local file path.
+     *
+     * @return bool
+     * @throws \coding_exception
+     */
+    protected function copy_file($remotefile, $localfile)
+    {
+        $result = copy($remotefile, $localfile);
+
+        if ($result) {
+            $this->log('copy_from_folder', 'Completed copy ' . $remotefile . ' to ' . $localfile);
```

Comment:
Should this say copy_from_url instead of  copy_from_folder ? (I'm not sure, asking)